fdbserver/workloads/KVStoreTest.actor.cpp

```diff
@@ -252,6 +303,204 @@ struct KVStoreTestWorkload : TestWorkload {
 
 WorkloadFactory<KVStoreTestWorkload> KVStoreTestWorkloadFactory("KVStoreTest");
 
+ACTOR Future<Void> testRangeReadClearRanges(KVStoreTestWorkload* workload, KVTest* ptest) {
+	/*
+	    - This test attempts to simulate a range read with various clear ranges
+	    - First we ask our old range read function to read a range of keys, lets say [a,m]
+	    - Then we randomly pick a subset of these keys, lets say [d,i) and designate this as our "clear range"
+	    - We then feed the range [a, m] to the new range read function and get the result
+	    - We then assert if the result contains all the keys in the range [a, m] except for those in [d, i)
+	*/
+	state KVTest& test = *ptest;
+	state Key key;
+	state int byteLimt = 1 << 30;
+	state int rowLimit = 1000;
+	state int count = 0;
+
+	while (true) {
+		// get the inital range of keys from the old read range function
+		state KeyRangeRef range = KeyRangeRef(key, LiteralStringRef("\xff\xff\xff\xff"));
+		state Standalone<RangeResultRef> kv = wait(test.store->readRange(range, rowLimit, byteLimt));
+		state Reference<ClearRangeResultWriter> resultWriter =
+		    makeReference<ClearRangeResultWriter>(rowLimit, byteLimt);
+
+		if (kv.size() == 0)
+			break;
+		// pick a random range to designate as our clear range
+		int rand1 = deterministicRandom()->randomInt(0, kv.size()),
+		    rand2 = deterministicRandom()->randomInt(0, kv.size());
+		state int clearStartIdx = std::min(rand1, rand2);
+		state int clearEndIdx = std::max(rand1, rand2);
+		// get the results from our new range read function with the range above
+		resultWriter->setClearRange(kv[clearStartIdx].key, kv[clearEndIdx].key);
+		wait(test.store->readRange(KeyRangeRef(kv[0].key, keyAfter(kv[kv.size() - 1].key)), resultWriter, rowLimit));
+		ASSERT(resultWriter->result.size() == kv.size() - (clearEndIdx - clearStartIdx));
+		int curIdx = 0;
+		// assert our range read function contains only the keys which are not part of the selected clear range
+		for (int i = 0; i < kv.size(); i++) {
+			if (i < clearStartIdx || i >= clearEndIdx) {
+				ASSERT(kv[i].key == resultWriter->result[curIdx].key);
+				ASSERT(kv[i].value == resultWriter->result[curIdx].value);
+				curIdx++;
+			}
+		}
+		ASSERT(curIdx == resultWriter->result.size());
+		count += resultWriter->result.size();
+		if (kv.size() < rowLimit)
+			break;
+		key = keyAfter(kv[kv.size() - 1].key);
+	}
+	TraceEvent("KVStoreRangeReadClearCount").detail("Count", count);
+
+	rowLimit = -1000;
+	key = LiteralStringRef("\xff\xff\xff\xff");
+	count = 0;
+
+	// same logic as above except we reverse the direction of the range read results (desc order)
```

Comment:
I think having a separate function for the forward range read and the reverse range read makes sense, that way if one breaks it is easier to see the issue. It also may be possible to make this a more generic function and parameterize it to be able to read in the forward or reverse direction.

Reply:
I took a look but this is not that much more beneficial. For clear ranges and such when going in reverse the logic for assertions and different. If I were to abstract this into another function it would just end up having a bunch of if-statements. Also in terms of failures since we use `ASSERT` if something fails we should know which direction is broken.